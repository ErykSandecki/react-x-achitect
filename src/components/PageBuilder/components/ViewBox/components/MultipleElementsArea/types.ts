// types
import { TElement, TRectCoordinates } from 'types';

export type TElements = {
  elementsCordinates: Array<{
    coordinates: TRectCoordinates;
    id: TElement['id'];
  }>;
  outline: TRectCoordinates;
};
