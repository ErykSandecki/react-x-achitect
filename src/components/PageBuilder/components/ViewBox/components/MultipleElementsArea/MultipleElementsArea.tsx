import { createPortal } from 'react-dom';
import { defer } from 'lodash';
import { FC, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

// components
import ClickableArea from '../ClickableArea/ClickableArea';
import Corners from '../Corners/Corners';

// core
import { useRefs } from 'pages/PageBuilderPage/core/RefsProvider';

// others
import { BASE_RECT } from 'shared';

// store
import {
  eventSelectorCreator,
  multipleSelectedElementsSelector,
  selectedElementsSelector,
} from 'store/pageBuilder/selectors';

// types
import { TElements } from './types';

// utils
import { getCoordinates } from './utils/getCoordinates';

const MultipleElementsArea: FC = () => {
  const selectedElements = useSelector(selectedElementsSelector);
  const isMultiple = useSelector(multipleSelectedElementsSelector);
  const { itemsRefs, overlayContainerRef } = useRefs();
  const isMultipleMoving = useSelector(
    eventSelectorCreator('isMultipleMoving'),
  ) as boolean;
  const [elements, setElements] = useState<TElements>({
    elementsCordinates: [],
    outline: BASE_RECT,
  });

  useEffect(() => {
    defer(() => {
      const coordinates = getCoordinates(
        isMultipleMoving,
        selectedElements,
        itemsRefs,
      );

      setElements(coordinates);
    });
  }, [isMultipleMoving, selectedElements]);

  if (!isMultiple) {
    return null;
  }

  return (
    <>
      <ClickableArea
        elementsCordinates={elements.elementsCordinates}
        rectCoordinates={elements.outline}
      />
      {!isMultipleMoving &&
        createPortal(
          <Corners rectCoordinates={elements.outline} />,
          overlayContainerRef.current,
        )}
    </>
  );
};

export default MultipleElementsArea;
