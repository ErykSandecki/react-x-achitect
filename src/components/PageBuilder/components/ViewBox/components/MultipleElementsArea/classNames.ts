export const className = 'MultipleElementsArea';

export const classNames = {
  [className]: className,
};
