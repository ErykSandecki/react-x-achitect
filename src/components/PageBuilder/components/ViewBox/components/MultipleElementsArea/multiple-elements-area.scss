@import '../../../../styles/z-indexes';

.MultipleElementsArea {
  &__shadow-element {
    position: absolute;
    z-index: $--rx-zindex-multiple-area;
  }
}
