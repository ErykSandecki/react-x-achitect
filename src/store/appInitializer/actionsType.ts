export const REDUCER_KEY = 'appInitializer';

export const APP_INIT = `${REDUCER_KEY}/APP_INIT`;
export const INIT_LANGUAGE = `${REDUCER_KEY}/INIT_LANGUAGE`;
export const INIT_LANGUAGE_SUCCESS = `${REDUCER_KEY}/INIT_LANGUAGE_SUCCESS`;
export const SET_IS_APP_LOADED = `${REDUCER_KEY}/SET_IS_APP_LOADED`;
export const SET_THEME = `${REDUCER_KEY}/SET_THEME`;
