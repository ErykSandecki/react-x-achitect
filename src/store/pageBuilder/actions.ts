// others
import {
  ADD_ELEMENT,
  SELECT_ELEMENT,
  UNSELECT_ELEMENT,
  SELECT_ELEMENTS,
  SET_AREA_COORDINATES,
  SET_ELEMENT_COORDINATES,
  SET_ELEMENTS_COORDINATES,
  UPDATE_EVENTS_STATUS,
  SET_ELEMENT_SIZES,
} from './actionsType';

// types
import {
  TAddELementAction,
  TAddELementActionPayload,
  TSelectElementAction,
  TUnselectElementAction,
  TSelectElementsAction,
  TSetAreaCoordinatesAction,
  TSetElementCoordinatesAction,
  TSetElementCoordinatesActionPayload,
  TSetElementsCoordinatesAction,
  TUpdateEventsStatusAction,
  TSetElementSizesAction,
  TSetElementSizesActionPayload,
} from './types';

export const addElement = (
  payload: TAddELementActionPayload,
): TAddELementAction => ({
  payload,
  type: ADD_ELEMENT,
});

export const selectElement = (
  payload: TSelectElementAction['payload'],
): TSelectElementAction => ({
  payload,
  type: SELECT_ELEMENT,
});

export const selectElements = (
  payload: TSelectElementsAction['payload'],
): TSelectElementsAction => ({
  payload,
  type: SELECT_ELEMENTS,
});

export const setAreCoordinates = (
  payload: TSetAreaCoordinatesAction['payload'],
): TSetAreaCoordinatesAction => ({
  payload,
  type: SET_AREA_COORDINATES,
});

export const setElementCoordinates = (
  id: TSetElementCoordinatesActionPayload['id'],
  positionAbsolute: TSetElementCoordinatesActionPayload['positionAbsolute'],
): TSetElementCoordinatesAction => ({
  payload: { id, positionAbsolute },
  type: SET_ELEMENT_COORDINATES,
});

export const setElementSizes = (
  baseCoordinates: TSetElementSizesActionPayload['baseCoordinates'],
  height: TSetElementSizesActionPayload['height'],
  id: TSetElementSizesActionPayload['id'],
  mouseCoordinates: TSetElementSizesActionPayload['mouseCoordinates'],
  width: TSetElementSizesActionPayload['width'],
): TSetElementSizesAction => ({
  payload: { baseCoordinates, height, id, mouseCoordinates, width },
  type: SET_ELEMENT_SIZES,
});

export const setElementsCoordinates = (
  payload: TSetElementsCoordinatesAction['payload'],
): TSetElementsCoordinatesAction => ({
  payload,
  type: SET_ELEMENTS_COORDINATES,
});

export const updateEventsStatus = (
  payload: TUpdateEventsStatusAction['payload'],
): TUpdateEventsStatusAction => ({
  payload,
  type: UPDATE_EVENTS_STATUS,
});

export const unselectElement = (
  payload: TUnselectElementAction['payload'],
): TUnselectElementAction => ({
  payload,
  type: UNSELECT_ELEMENT,
});
