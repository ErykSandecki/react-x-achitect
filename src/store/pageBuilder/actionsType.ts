export const REDUCER_KEY = 'pageBuilder';

export const ADD_ELEMENT = `${REDUCER_KEY}/ADD_ELEMENT`;

export const ROTATE_ELEMENT = `${REDUCER_KEY}/ROTATE_ELEMENT`;
export const SELECT_ELEMENT = `${REDUCER_KEY}/SELECT_ELEMENT`;
export const SELECT_ELEMENTS = `${REDUCER_KEY}/SELECT_ELEMENTS`;
export const SET_AREA_COORDINATES = `${REDUCER_KEY}/SET_AREA_COORDINATES`;
export const SET_ELEMENT_COORDINATES = `${REDUCER_KEY}/SET_ELEMENT_COORDINATES`;
export const SET_ELEMENT_SIZES = `${REDUCER_KEY}/SET_ELEMENT_SIZES`;
export const SET_ELEMENTS_COORDINATES = `${REDUCER_KEY}/SET_ELEMENTS_COORDINATES`;
export const UPDATE_EVENTS_STATUS = `${REDUCER_KEY}/UPDATE_EVENTS_STATUS`;
export const UNSELECT_ELEMENT = `${REDUCER_KEY}/UNSELECT_ELEMENT`;
