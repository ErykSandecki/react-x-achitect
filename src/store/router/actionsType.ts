export const REDUCER_KEY = 'router';

export const HISTORY_CHANGED = `${REDUCER_KEY}/HISTORY_CHANGED`;
export const PUSH = `${REDUCER_KEY}/PUSH`;
export const REDIRECT_TO = `${REDUCER_KEY}/REDIRECT_TO`;
