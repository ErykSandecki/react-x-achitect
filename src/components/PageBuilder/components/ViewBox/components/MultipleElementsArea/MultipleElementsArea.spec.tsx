import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import MultipleElementsArea from './MultipleElementsArea';

// mocks
import {
  pageBuilderStateMock,
  selectedElementMock,
} from 'test/mocks/reducer/pageBuilderMock';

// others
import { REDUCER_KEY as PAGE_BUILDER } from 'store/pageBuilder/actionsType';

// store
import { configureStore } from 'store/store';

const stateMock = {
  [PAGE_BUILDER]: {
    ...pageBuilderStateMock[PAGE_BUILDER],
    selectedElements: {
      [selectedElementMock.id]: selectedElementMock,
      ['2']: {
        ...selectedElementMock,
        id: '2',
      },
    },
  },
};

describe('MultipleElementsArea snapshots', () => {
  it('should render MultipleElementsArea', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <MultipleElementsArea />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not render MultipleElementsArea', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...pageBuilderStateMock[PAGE_BUILDER],
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <MultipleElementsArea />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
