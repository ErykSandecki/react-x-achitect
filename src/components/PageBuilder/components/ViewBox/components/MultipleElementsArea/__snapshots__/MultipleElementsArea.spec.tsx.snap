// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleElementsArea snapshots should not render MultipleElementsArea 1`] = `<DocumentFragment />`;

exports[`MultipleElementsArea snapshots should render MultipleElementsArea 1`] = `
<DocumentFragment>
  <svg
    class="ClickableArea ClickableArea--light"
    style="height: 1.5px; left: -0.75px; top: -0.75px; width: 1.5px;"
  >
    <path
      class="ClickableArea__outline ClickableArea__outline--light"
      d="M 0,0 H 1.5 V 1.5 H 0 V 0"
      stroke-width="1.5"
    />
  </svg>
  <svg
    class="Corners Corners--light Corners--increase-zindex Corners--increase-zindex--light"
    style="height: 1.5px; left: -0.75px; top: -0.75px; width: 1.5px;"
  >
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="-3"
      y="-3"
    />
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="-3"
      y="-1.5"
    />
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="-1.5"
      y="-3"
    />
    <rect
      height="6"
      rx="1.5"
      width="6"
      x="-1.5"
      y="-1.5"
    />
  </svg>
</DocumentFragment>
`;
